bpf:nat: Restore ORG NAT entry if it's not found

gyutaeb opened the pull request:
<!-- Description of change -->
# Problem Statement

- In client-to-service communication, when traffic is load-balanced to a remote endpoint, the remote endpoint sends an RST due to incorrect SNAT(new source port). As a result, the socket and session are terminated.
- The same issue also occurs in pod-to-external communication.
- For more details, refer to #36572 

# Root Cause

- This issue occurs when the original NAT entry is evicted from the SNAT_MAPPING map due to the LRU policy. As a result, even for a flow that has already undergone SNAT, a new source port is assigned. The remote endpoint then terminates the session due to the unexpected source port.

# Solution Summary

- Recreating the deleted original entry in snat_v4_rev_nat_handle_mapping().
![image](https://github.com/user-attachments/assets/70e2bdf1-6c61-4b23-bba5-6f586db1ccef)

# Impact Assessment

This patch has a substantial impact on connection reliability. In testing, it reduced the connection failure rate from 1–10% to virtually 0%, effectively eliminating the issue.

![image](https://github.com/user-attachments/assets/9c5dc5df-015e-492c-9b97-6d89219d5f3f)

# Issue Details

- snat_v4_rev_nat_handle_mapping() correctly finds the reverse entry.
- When the egress packet reaches snat_v4_nat_handle_mapping(), the original entry has already been evicted by LRU.
- Since the original entry is missing, snat_v4_new_mapping() assigns a new source port for NAT.
- If this happens, the endpoint socket detects a port change and sends an RST packet.

# eBPF Unit Test Flow

- The original packet gets ReSNATed when the entry is deleted.
- The reply packet restores the Original NAT entry if it is deleted.
- The original pakcket does not get ReSNATed. Because the Original NAT entry is restored.

----

Network failures can occur due to LRU eviction in the BPF map. This happens because NAT entries are evicted during packet transmission. Inspired by #35304, I developed this PR to help mitigate this issue. This PR restores deleted original NAT entries in `snat_v4_rev_nat_handle_mapping()`. The idea was initially proposed at https://github.com/cilium/cilium/pull/36631#issuecomment-2650776896, and this PR serves as its actual implementation.
Thanks to @julianwiedmann @sugangli

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -406,6 +409,29 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+			ostate.to_saddr = tuple->daddr;
+			ostate.to_sport = tuple->dport;
+			ostate.common.needs_ct = (*state)->common.needs_ct;
+			ostate.common.created = bpf_mono_now();
```
I wonder whether it's more correct to take the timestamp from `state->common.created`. Same applies to the [previous patch](https://github.com/cilium/cilium/pull/35304/commits/b4cabeed66fb238d73f776215710462abfd76399#diff-ef710180f228e8f6bbad249e2670dbc42a54f45da3bbc4c5aebbb1f3865979e9R331). We set them to the [same value](https://github.com/cilium/cilium/blob/7e38bb538b30498a4ebc39b8013ea2fd7c73b37e/bpf/lib/nat.h#L271) when creating both, and here we are trying to restore the old one, not creating a new.

gyutaeb replied:
@gentoo-root I've also put a lot of thought into this code. However, setting the restored entry's timestamp to `bpf_mono_now()` has the advantage of clearly indicating whether an entry is newly created or restored. Setting the [same value](https://github.com/cilium/cilium/blob/7e38bb538b30498a4ebc39b8013ea2fd7c73b37e/bpf/lib/nat.h#L271) is appropriate because both the original entry and the reverse entry are handled within the same context. (although not perfectly the same time, it is close enough.)

gentoo-root commented on bpf/lib/nat.h:
```diff
@@ -406,6 +409,29 @@ snat_v4_rev_nat_handle_mapping(struct __ctx_buff *ctx,
+		/* Check for the original SNAT entry. If it is missing (e.g. due to LRU
+		 * eviction), it must be restored before returning.
+		 */
+		org_tuple.saddr = (*state)->to_daddr;
```
Nit: let's name it orig_tuple (it's a better abbreviation of "original") or otuple (like in the existing code).

gyutaeb replied:
@gentoo-root Thank you for the detailed review :) I will update the name accordingly.

gyutaeb replied:
@gentoo-root Done. I used `otuple`.

gentoo-root reviewed (commented):
Looks good to me (modulo minor comments). Would be good if you could add a test.

gyutaeb commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -1177,3 +1177,287 @@ int nodeport_nat_fwd_restore_reply_check(const struct __ctx_buff *ctx)
+  */
+PKTGEN("tc", "tc_nodeport_nat_fwd_original_renated")
+int nodeport_nat_fwd_original_reanted_pktgen(struct __ctx_buff *ctx)
+{
```
@gentoo-root Thank you for waiting. I tested the ReSNATed behavior when the Original NAT entry was deleted. I also confirmed that with this patch, ReSNATed no longer occurs after the Original NAT entry is removed, and communication remains stable.

gentoo-root commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -1177,3 +1177,287 @@ int nodeport_nat_fwd_restore_reply_check(const struct __ctx_buff *ctx)
+  * Test that source port is changed when the Original NAT entry is deleted.(ReSNATed)
+  */
+PKTGEN("tc", "tc_nodeport_nat_fwd_original_renated")
+int nodeport_nat_fwd_original_reanted_pktgen(struct __ctx_buff *ctx)
```
Nit: typo reanted -> renated.

gyutaeb replied:
@gentoo-root Oops, I made a mistake. I've applied the fix. Thanks! 😊

gentoo-root commented on bpf/tests/tc_nodeport_lb4_nat_lb.c:
```diff
@@ -1177,3 +1177,287 @@ int nodeport_nat_fwd_restore_reply_check(const struct __ctx_buff *ctx)
+	otuple.dport = BACKEND_PORT;
+
+	/* Delete the Original NAT entry*/
+	if IS_ERR(map_delete_elem(&SNAT_MAPPING_IPV4, &otuple))
```
Am I right to assume that this test relies on the state that remains after the previous tests? That's not the best approach, because tests can be started individually, and also I'm not sure that we have guarantees on a particular order. Although I see that all tests in this file do this (in particular, storing nat_source_port for the next test), so maybe it's OK to keep this (anti)pattern...

If that's the case, do we actually need those lb4_add_service etc. calls above?

gyutaeb replied:
@gentoo-root Thanks for the review!
> Am I right to assume that this test relies on the state that remains after the previous tests? That's not the best approach, because tests can be started individually, and also I'm not sure that we have guarantees on a particular order. Although I see that all tests in this file do this (in particular, storing nat_source_port for the next test), so maybe it's OK to keep this (anti)pattern...

You're right. I agree with your thoughts. However, I followed the existing structure because it was already interdependent across tests.

> If that's the case, do we actually need those lb4_add_service etc. calls above?

In fact, lb4_add_service is not strictly necessary. However, I tried to follow the explicit approach from #35304, where lb4_add_service was explicitly called. https://github.com/cilium/cilium/blob/50c98730df93d4eb913bcc7cc40267923d9db8e0/bpf/tests/tc_nodeport_lb4_nat_lb.c#L1071-L1087

gentoo-root reviewed (approved):
Sorry for the delay. Just one non-blocking comment/question on the test.

gyutaeb committed:
bpf:nat: Restore ORG NAT entry if it's not found

ORG NAT entry could be removed by LRU eviction. This commit restore the
ORG NAT entry from REV NAT entry and tuple in
snat_v4_rev_nat_handle_mapping(). Thus, it can mitigate network failures
caused by LRU eviction.

Relates: #35304
Signed-off-by: Gyutae Bae <gyu.8ae@gmail.com>

gyutaeb commented on bpf/lib/nat.h:
```diff
@@ -1361,6 +1391,36 @@ snat_v6_rev_nat_handle_mapping(struct __ctx_buff *ctx,
 {
 	*state = snat_v6_lookup(tuple);
 
+	if (*state) {
+		struct ipv6_nat_entry *lookup_result;
+		struct ipv6_nat_entry ostate;
+		struct ipv6_ct_tuple otuple = {};
+		int ret;
+
+		/* Check for the original SNAT entry. If it is missing (e.g. due to LRU
+		 * eviction), it must be restored before returning.
+		 */
+		otuple.saddr = (*state)->to_daddr;
+		otuple.sport = (*state)->to_dport;
+		otuple.daddr = tuple->saddr;
+		otuple.dport = tuple->sport;
+		otuple.nexthdr = tuple->nexthdr;
+		otuple.flags = TUPLE_F_OUT;
+
+		lookup_result = snat_v6_lookup(&otuple);
+		if (!lookup_result) {
+			memset(&ostate, 0, sizeof(ostate));
+			ostate.to_saddr = tuple->daddr;
+			ostate.to_sport = tuple->dport;
+			ostate.common.needs_ct = (*state)->common.needs_ct;
+			ostate.common.created = bpf_mono_now();
+
+			ret = __snat_create(&cilium_snat_v6_external, &otuple, &ostate);
+			if (ret < 0)
+				return DROP_NAT_NO_MAPPING;
+		}
+	}
+
```
The IPv6 path follows the same principle as IPv4.
